### PR TITLE
deduplicate the 'isSomeFunction' equivalent

### PR DESCRIPTION
**Pull request by UplinkCoder**
Since the code is now used multiple times in `__traits` let's dedup it.

### CODE REVIEW

File: src/ddmd/traits.d
Comment (rainers) [inline src/ddmd/traits.d]: IMO a better abstraction is `toTypeFunction(RootObject o)` and passing the validated parameter into this function. That would allow it not being a nested function and used elsewhere, too. 

Even better, I'd split it into two overloads, one for Type, and one for Dsymbol, as you now have widen the allowed parameters to traits that support types only so far.

Comment (UplinkCoder) [inline src/ddmd/traits.d]: the reason I did not do this is because I'd have to import `rootobject`

Comment (UplinkCoder) [inline src/ddmd/traits.d]: also let's generalize when/if needed and not on speculation.

**Comment by UplinkCoder** [comment]
travis fails because of connection error

**Comment by wilzbach** [comment]
@uplinkcoder: restarted and created another card for it https://trello.com/c/aM5r3Bio/320-use-more-download-fallbacks-in-the-installer-script

File: src/ddmd/traits.d
Comment (UplinkCoder) [inline src/ddmd/traits.d]: @rainers do you agree ?

Comment (rainers) [inline src/ddmd/traits.d]: > do you agree ?

Mostly. The diff didn't show that there's also support for function symbols in the other traits.

You should remove the handling for FuncDeclaration in traits getLinkage, getFunctionVariadicStyle and getParameterStorageClasses, though. It is duplicated now.

> the reason I did not do this is because I'd have to import rootobject

That's not really a good excuse. You can also use typeof(args[0]) to avoid an import... ;-)

**Comment by UplinkCoder** [comment]
@rainers do you like it better now ?

**Comment by UplinkCoder** [comment]
@WalterBright any objections ?

File: src/ddmd/traits.d
Comment (WalterBright) [inline src/ddmd/traits.d]: No documentation saying what function does.

**Comment by WalterBright** [comment]
LGTM provided the function is documented

**Comment by WalterBright** [comment]
>use typeof(args[0]) to avoid an import... ;-)

Not a bad idea.

File: src/ddmd/traits.d
Comment (WalterBright) [inline src/ddmd/traits.d]: Please use Ddoc style documentation. It this case it means having a `Params:` section and a `Returns:` section. This is the standard for new functions.

File: src/ddmd/traits.d
Comment (WalterBright) [inline src/ddmd/traits.d]: omit trailing period in middle of sentence

File: src/ddmd/traits.d
Comment (WalterBright) [inline src/ddmd/traits.d]: `passed verbatim` has no meaning in regards to code.

File: src/ddmd/traits.d
Comment (wilzbach) [inline src/ddmd/traits.d]: @UplinkCoder don't forget that we already publish the documentation of ddmd on dlang.org:
- http://dlang.org/phobos-prerelease/ddmd_mars.html (Ddoc)
- http://dlang.org/library-prerelease/ddmd/mars.html (Ddox)
and we intend to publish it on the main page very soon (e.g. once #6771 is merged).

This is also why it can be very helpful for both the contributor and the reviewers to preview the style changes at Vladimir's excellent DAutoTest, e.g.

http://dtest.dlang.io/artifact/website-72ba79b73f635fb08bdfe69d6826833c1c4c2ee0-6f020744a26d63533038dc39e44ae752/web/library-prerelease/ddmd/traits.html

(in this case it's a function within a function and thus _not_ publicly accessible)

File: src/ddmd/traits.d
Comment (WalterBright) [inline src/ddmd/traits.d]: When using periods, please do not put them in the middle of sentences. Replace it with a `,`

**Comment by UplinkCoder** [comment]
@WalterBright addressed.

File: src/ddmd/traits.d
Comment (UplinkCoder) [inline src/ddmd/traits.d]: @WalterBright do you like it better now ?

File: src/ddmd/traits.d
Comment (WalterBright) [inline src/ddmd/traits.d]: Please read the Ddoc documentation I provided a link for. Note the `o:` in the above line. This is not how it's done and Ddoc will not work with it. It's `o=`.

**Comment by JinShil** [comment]
@UplinkCoder Are you going to take this across the finish line?

File: src/ddmd/traits.d
Comment (JinShil) [inline src/ddmd/traits.d]: declararion --> declaration

File: src/ddmd/traits.d
Comment (JinShil) [inline src/ddmd/traits.d]: I recommend a different name.  Perhaps `getTypeFunction`.

File: src/ddmd/traits.d
Comment (JinShil) [inline src/ddmd/traits.d]: I offer you this documentation header:

```D
/**
 Gets the `TypeFunction` for `o`
 Params:
    o   = the AST node to get the `TypeFunction` for
    fdp = optional pointer to a function declararion, to be set 
          if `o` is a function declarartion
 Returns: 
    a `TypeFunction` if `o` is a declaration for a delegate, function,
    or avariable of those types.  Otherwise, `null`
*/
```

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/ddmd/traits.d
Comment (stefan-koch-sociomantic) [reply src/ddmd/traits.d]: Neat Thanks!

File: src/ddmd/traits.d
Comment (BBasile) [inline src/ddmd/traits.d]: `if o`

File: src/dmd/traits.d
Comment (JinShil) [inline src/dmd/traits.d]: All other DDoc headers in DMD seem to follow this convention:

```D
/*************************************************************
 * Whatever
 * Whatever
 */
```

Please remove the extra space between the DDoc header and the function declaration.

I cannot, in good conscience, approve a function name like `traitsFuncArg`.  Please use a more descriptive name like `getFuncType` or something along those lines (Yes, I know my own names have been poor).

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/dmd/traits.d
Comment (wilzbach) [reply src/dmd/traits.d]: > Please remove the extra space between the DDoc header and the function declaration.

Hmm the above is the Phobos style:
Compare with e.g. https://github.com/dlang/phobos/blob/master/std/algorithm/comparison.d

If you really want to stick with this style at DMD, then this should be documented in a DMD-specific style guide?

Comment (JinShil) [reply src/dmd/traits.d]: > Hmm the above is the Phobos style.

I see.

> If you really want to stick with this style at DMD, then this should be documented in a DMD-specific style guide.

Agreed.  

Aside from the name then, LGTM.

Comment (JinShil) [reply src/dmd/traits.d]: Also, consider making this a non-member function instead of static.  Refer to [@WalterBright's comment](https://github.com/dlang/dmd/pull/7405#pullrequestreview-82322025)

Comment (jacob-carlborg) [reply src/dmd/traits.d]: Phobos has a mixed style when it comes to comments [1] [2] [3].

[1] https://github.com/dlang/phobos/blob/master/std/string.d#L2194-L2205
[2] https://github.com/dlang/phobos/blob/master/std/string.d#L2254-L2267
[3] https://github.com/dlang/phobos/blob/master/std/string.d#L2332-L2358

Comment (andralex) [reply src/dmd/traits.d]: Let's use https://github.com/dlang/phobos/blob/master/std/string.d#L2254-L2267 for new code, thanks.

**Comment by JinShil** [comment]
This PR needs to have the reviewers' concerns addressed and rebased.

This PR is nice, and normally I would revive such a PR if the author was no longer active.  I would also revive it if it were an important bug fix or something of higher priority.  This PR, however, is just a refactoring, so I'm going to put it in the [Phantom Zone](http://forum.dlang.org/post/ouuutodvhmnghzbeoqen@forum.dlang.org) and close it for now.

**Comment by JinShil** [comment]
Resubmitted at #7749

**Pull request closed without merging**
